### DISCUSSION TITLE
How do you pass payload to dynamic routes in nuxt generate?

### DISCUSSION THREAD

**Discussion by surgiie** [nuxt/nuxt#22006] [Questions]
In nuxt 2, you were able to specify `payload` in a route object during a `nuxt generate` to provide the data for the static html file:

https://v2.nuxt.com/docs/configuration-glossary/configuration-generate/#speeding-up-dynamic-route-generation-with-payload

I dont see any docs on how to accomplish this in nuxt 3, passing this payload during a generate is not rendering this data as it was before.

Am i missing something?

**Comment by prakashpie** [comment]
@surgiie in **Nuxt 3**,  you don't have to pass any payload, which is very useful when you want to view those pages in dev mode. Previously, in order to see those generated pages, we had to generate all pages and then view them using a local server.

### Now here is your answer:

## Step 1 : creating the [slug].vue file

```vue
// pages/[slug].vue
<template>
  <common-generic-template :pageData="pageData" :breadcrumb="breadcrumb" />
</template>
<script setup lang="ts">
definePageMeta({
  layout: 'default',
})

const route = useRoute()
const config = useRuntimeConfig()
const { data: pageData, error } = await useAsyncData(
  `${route.params.slug}`,
  () =>
    $fetch(`${config.WORDPRESS_URL}pages`, {
      params: {
        slug: route.params.slug,
      },
    }).catch((error) => error.data),
  {
    transform: (resData) => {
      return resData.length ? resData[0] : {}
    },
  },
)

const { setHeadByPageData } = useSeoHead()
setHeadByPageData(pageData.value)

const breadcrumb = [
  {
    text: 'Home',
    to: '/',
  },
  {
    text: pageData.value.title.rendered,
    active: true,
  },
]
</script>

```

Here `pageData` will act as the payload. Also, make sure you add a unique key to the `useAsyncData` function.
`${route.params.slug}` serves as the unique key for each page. Without it, when you navigate from one Nuxt link to another, the page data won't change. 
Please note that the `transform` function is used because I am generating pages using the WordPress API, which returns an array, and I transform it into an object.

I am using `useSeoHead` (a local composable for transforming and extracting SEO information) to set the title, description, and other SEO-related aspects using the `pageData` object. However, you can also use the built-in `useHead` composable.

## Step 2 : Setting up the API call to get dynamic slug

```ts
// api/index.ts

export async function fetchGeneric() {
  return await fetch(`${process.env.WORDPRESS_URL}pages/`).then((res) =>
 res.json())
}

```

## Step 3 : Setting up the nitro routes configuration

```ts
// nuxt.config.ts

import { fetchGeneric } from './api'
```

```ts
// nuxt.config.ts

nitro: {
    hooks: {
      async 'prerender:routes'(routes) {
        const allRoutes = []

        // Generic pages routes | only required slug information
        const genericPages = await fetchGeneric()
        if (genericPages) {
          const genericRoutes = genericPages.map((page: any) => {
            return `/${page.slug}/`
          })
          allRoutes.push(...genericRoutes)
        }

        if (allRoutes.length) {
          for (const route of allRoutes) {
            routes.add(route)
          }
        }
      },
    },
  },
````
Voila 🤩, You can now generate static pages without any issue.
You can always reach out to [me](mailto:prakash.codepie@gmail.com) if you need any additional help.

**Reply by surgiie** [reply]
@prakashpie

I have an odd workflow where we run `nuxt generate` in background processes isolated from our application's deployments and we use files for payload. 

We essentially do something like  `nuxt generate —file=/some/file.json`

In our nuxt 2 application we have a nuxt module where we would parse this option flag and load the file and register the dynamic route:

```js

 nuxt.hook("generate:extendRoutes", function (routes) {
    // parse --file options from process.argv
    // load data from file then do something like this for each file passed:
    routes.push({
        route: "/dynamic/url/from/file/data",
        payload: {
            content: data
        },
    })
 })

```

Then within our page component we were able to just return the payload in asyncData ( no need to fetch/parse files, it was already provided)

```js
async asyncData({payload}) {
    return {
        content: payload
    }
},
```


This method worked great in nuxt 2 and i would prefer the ability to continue doing so in nuxt 3.

**Reply by surgiie** [reply]
I have a work around currently by parsing the file 2 times, once in the `prerender:routes` hook (to build the slug) and once in `asyncData` (to return the payload). Im able to compute the file path in `asyncData` from the route. It works but i would prefer the ability to pass in the payload since it was already parsed and processed during the `prerender:routes` hook which then reduces this down to 1 file load instead of 2.

**Reply by Hebilicious** [reply]
You're gonna have to stick with the workarounds for a little longer, there's a proposal to add more flexibility around that area https://github.com/nuxt/nuxt/issues/21855. Would that cover your use-case ?

**Reply by surgiie** [reply]
@Hebilicious Yeah this looks like the flexibility I need. For now, a work around is fine.

**Reply by AndreTheSpire** [reply]
hello i just also trying out the full static website on nuxt 3 by doing same as what u said but i very confused why my page keep fetching data even thought i already generate the payload.

**Comment by gavinthomas-valtech** [comment]
i'm also trying to generate dynamic routes. Where's the nuxt documentation on having to use `useAsyncData` with a unique key to generate routes? struggling to find examples of this

**Reply by prakashpie** [reply]
I have done this already in Nuxt 3. Let me know so that i can help you.

**Reply by AndreTheSpire** [reply]
@prakashpie may i ask if the way u do the solution will still fetch the useAsyncData every time to load the slug.vue page?  cause i already try your solution and check nuxtApp.payload and the fetchdata is there but my page still fetch the useAsyncData every time i go to the page.

**Comment by prakashpie** [comment]
```vue
<template>
 //your page html code
</template>

<script setup lang="ts">
definePageMeta({
  layout: 'default',
})

const route = useRoute()
const config = useRuntimeConfig()

const { data: pageData, error } = await useAsyncData(
  `${route.params.slug}`, // this will be use as unique key for each slug
  () =>
    $fetch(`${config.WORDPRESS_URL}pages`, { // calling wordpress api but you can use your own
      params: {
        slug: route.params.slug,
      },
    }),
  {
    transform: (resData: any) => {
      return resData?.length ? resData[0] : {}
    },
  },
)
```
Hope this code snippet will solve your issue @AndreTheSpire

**Reply by AndreTheSpire** [reply]
@prakashpie Isnt that same as the step 1 solution that u sent above? I already try that but unfortunately mine for some reason still fetch the data when i load the page. But what more i confused is when i check the value of NuxtApp.payload.data there an object of the data fetched in generate but its still keep fetch data everytime i go to the slug page. Do maybe know what the reason?

**Comment by prakashpie** [comment]
@AndreTheSpire are you using 'nuxt generate'? If so, please ensure that you've configured all possible dynamic URLs in the Nitro config so that each URL generates a unique key payload.

Consider an example: on a website like https://www.rummyprime.com/blog/, there is no extra calls occur when navigating to any blogs.

Additionally, you can send me a Google Meet invite at prakash.codepie@gmail.com so that I can provide on-the-go assistance.

**Reply by AndreTheSpire** [reply]
@prakashpie tbh i using npx nuxi generate from nuxt 3 static documentation and for the dynamic url i already setting it in nitro config like below 
```vue
  hooks: {
    async 'nitro:config'(nitroConfig) {
      nitroConfig.prerender = nitroConfig.prerender || {};
      nitroConfig.prerender.routes = nitroConfig.prerender.routes || [];

      if (nitroConfig.dev) {
        return;
      }

      try {
        const response = await fetch(url);
          if (response.ok) {
            const news: News[] = await response.json();
            // console.log(news);
            const routes2 = news.map((news) => `/news/${news.slug}`);
            nitroConfig.prerender.routes.push(...routes2);
          } else {
            console.error('Error fetching data:', response.statusText);
          }
         
      } catch (error) {
        console.error('Error fetching data:', error);
      }
    },
  },
``` 
i can see the page being already generated in /.output/public but the payload just dont seems to save the data but strangely when i call NuxtApp.payload in the page its have the value of the fetch data but its still fetching even thought the payload have the value. is there something that i need to configurate more?
for more information i tried to see the payload using
``
const nuxtApp = useNuxtApp();
if (nuxtApp.payload) {
  console.log(nuxtApp.payload);
}
``
and the output of the payload is like this
![image (3)](https://github.com/nuxt/nuxt/assets/64310095/3b16657a-19e3-407f-86da-1110133e92f9)
as u can see the data have the gereja-berpolitik as the key of the fetched object

**Comment by prakashpie** [comment]
Hi everyone, please find the GitHub repo example for generating dynamic routes:

[https://github.com/codepie-io/nuxt3-dynamic-routes](https://github.com/codepie-io/nuxt3-dynamic-routes)

**Reply by surgiie** [reply]
Thanks for the example. While this works for alot, it doesnt work for my use case i mentioned above. IMO you should be able to still pass in the payload as you were able to in nuxt 2. This method requires you to fetch the data in `asyncData` while in nuxt 2, if you had previously had payload in the nuxt hook, you could pass it in. Not sure what to do other than working around this problem as i mentioned above.

**Reply by prakashpie** [reply]
Hi @surgiie ,

Thank you for clarifying your expectations. The payload feature was not available in Nuxt versions below v3.10.xx. However, it is now available in Nuxt v3.10 as an experimental feature. I have updated the code for both scenarios, with payloads and without payloads.

Please check out the repository at:

[nuxt3-dynamic-routes](https://github.com/codepie-io/nuxt3-dynamic-routes)

and let me know your thoughts on this.

I hope it will be helpful for you.

**Reply by surgiie** [reply]
> Hi @surgiie ,
> 
> 
> 
> Thank you for clarifying your expectations. The payload feature was not available in Nuxt versions below v3.10.xx. However, it is now available in Nuxt v3.10 as an experimental feature. I have updated the code for both scenarios, with payloads and without payloads.
> 
> 
> 
> Please check out the repository at:
> 
> 
> 
> [nuxt3-dynamic-routes](https://github.com/codepie-io/nuxt3-dynamic-routes)
> 
> 
> 
> and let me know your thoughts on this.
> 
> 
> 
> I hope it will be helpful for you.
> 
> 

Oh nice i didnt realize there was an experimental option for payload. I will look into it! Thanks.

**Comment by Aravinda93** [comment]
@prakashpie 

[How to pre-render all the routes in Nuxt 3 application .output automatically (without hard coding and specifying them specifically)?](https://stackoverflow.com/questions/78658283/how-to-pre-render-all-the-routes-in-nuxt-3-application-output-automatically-wi) Is there a way to pre-render all the routes automatically rather than defining the individual routes? I don't wish to add a method to identify the routes, but rather try to find a way where `nuxt generate` can automatically figure out the routes and pre-render them into the `.output/public`.

**Reply by prakashpie** [reply]
Hello @Aravinda93 ,  

It is possible if crawlLinks are enabled, though at least one page should have links to other pages. Think of footer links, which contain many links that Nitro can easily crawl. However, this approach may not work in certain scenarios.

For example, consider a blog page with more than 10 blogs where you render 5 blogs per page or use lazy load on scroll. In this case, if the blog index page link is added to the footer, header, or anywhere on the page, it will find 5 individual pages along with the blog index page. Unfortunately, it will not render the other 10 pages because Nitro does not know that those additional 5 blogs exist.

[https://nuxt.com/docs/getting-started/prerendering#crawl-based-pre-rendering](https://nuxt.com/docs/getting-started/prerendering#crawl-based-pre-rendering)

` nitro: {
    prerender: {
      crawlLinks: true
    },
  },`

**Comment by Aravinda93** [comment]
@prakashpie 

Thanks a lot for the response. I tried this but it's not working either:

```
export default {
  modules: ["@nuxtjs/tailwindcss", "unplugin-fonts/nuxt", "@nuxtjs/i18n"],

  ssr: false,

  experimental: {
    inlineRouteRules: true,
  },

  nitro: { prerender: { crawlLinks: true } },

  hooks: {},
}
```

I have the `/layouts/default.vue` where I have all the links to various pages enabled such as:
```
<li>
  <NuxtLink to="/page1"> Page-1 </NuxtLink>
</li>

<li>
  <NuxtLink to="/page2"> Page-2 </NuxtLink>
</li>
```

Not sure why all routes are pre-rendered already automatically. Please let me know if anything else I can try.

**Reply by prakashpie** [reply]
Hi @Aravinda93 
Please change the ssr false to true. then try again. Let me know if the problem still exists.

**Reply by Aravinda93** [reply]
@prakashpie I tried to run the `npm run generate` by setting the `ssr:true` but I ran into the error:

```
 ERROR  Cannot read properties of undefined (reading 'userAgent')                                                                                    10:03:14 AM

  at .nuxt/prerender/chunks/build/XMLEditor-CNiUFJWC.mjs:348:32
  at .nuxt/prerender/chunks/build/XMLEditor-CNiUFJWC.mjs:346:24
  at .nuxt/prerender/chunks/build/XMLEditor-CNiUFJWC.mjs:347:7
```

I believe it comes from the `Codemirror` library which I am using within my application. By checking a few answers I tried adding the `process.client` check to run this only on the client side but still running into this error. I believe it's coming from the Codemirror library internally not from within my code.

Is there any way we can pre-render the routes with `ssr:false`? Because I believe it was working well before. Is there anything to do with `nuxt/vue/node/npm` version updation or changes?

**Reply by prakashpie** [reply]
HI @Aravinda93 , If possible please connect me on [prakash.codepie@gmail.com](mailto:prakash.codepie@gmail.com). I will personally look into your project.

**Reply by Aravinda93** [reply]
Hello @prakashpie 

Many many thanks for offering the email and looking into the code.
I had a discussion with my lead and he agreed to include the required routes as a part of the hooks, instead of Nuxt 3 automatically looking for the routes. Hence with `ssr:false`, i added this to my `nuxt-config.js`:

```
  hooks: {
    "prerender:routes"({ routes }) {
      //Function to add routes for pre-render
      routes.add("/page1")
      routes.add("/page2")
      routes.add("/about")
    },
  },
```

So now its picking up and adding the pre-rendered routes to my `.output/public`. 

Thanks a lot again for offering the help, have a nice day :)
